crates/matrix-sdk-ui/src/timeline/controller/aggregations.rs

```diff
@@ -0,0 +1,429 @@
+    related_events: HashMap<TimelineEventItemId, Vec<Aggregation>>,
+
+    /// Mapping of a related event identifier to its target.
+    inverted_map: HashMap<TimelineEventItemId, TimelineEventItemId>,
+}
+
+impl Aggregations {
+    /// Clear all the known aggregations from all the mappings.
+    pub fn clear(&mut self) {
+        self.related_events.clear();
+        self.inverted_map.clear();
+    }
+
+    /// Add a given aggregation that relates to the [`TimelineItemContent`]
+    /// identified by the given [`TimelineEventItemId`].
+    pub fn add(&mut self, related_to: TimelineEventItemId, aggregation: Aggregation) {
+        self.inverted_map.insert(aggregation.own_id.clone(), related_to.clone());
+        self.related_events.entry(related_to).or_default().push(aggregation);
+    }
+
+    /// Is the given id one for a known aggregation to another event?
+    ///
+    /// If so, returns the target event identifier as well as the aggregation.
+    pub fn try_remove_aggregation(
+        &mut self,
+        aggregation_id: &TimelineEventItemId,
+    ) -> Option<(&TimelineEventItemId, Aggregation)> {
+        let found = self.inverted_map.get(aggregation_id)?;
+
+        // Find and remove the aggregation in the other mapping.
+        let aggregation = if let Some(aggregations) = self.related_events.get_mut(found) {
+            let removed = aggregations
+                .iter()
+                .position(|agg| agg.own_id == *aggregation_id)
+                .map(|idx| aggregations.remove(idx));
+
+            // If this was the last aggregation, remove the entry in the `related_events`
+            // mapping.
+            if aggregations.is_empty() {
+                self.related_events.remove(found);
+            }
+
+            removed
+        } else {
+            None
+        };
+
+        if aggregation.is_none() {
+            warn!("incorrect internal state: {aggregation_id:?} was present in the inverted map, not in related-to map.");
+        }
+
+        Some((found, aggregation?))
+    }
+
+    /// Apply all the aggregations to a [`TimelineItemContent`].
+    ///
+    /// Will return an error at the first aggregation that couldn't be applied;
+    /// see [`Aggregation::apply`] which explains under which conditions it can
+    /// happen.
+    pub fn apply(
+        &self,
+        item_id: &TimelineEventItemId,
+        content: &mut TimelineItemContent,
+    ) -> Result<(), AggregationError> {
+        let Some(aggregations) = self.related_events.get(item_id) else {
+            return Ok(());
+        };
+        for a in aggregations {
+            a.apply(content)?;
+        }
+        Ok(())
+    }
+
+    /// Mark a target event as being sent (i.e. it transitions from an local
+    /// transaction id to its remote event id counterpart), by updating the
+    /// internal mappings.
+    pub fn mark_target_as_sent(&mut self, txn_id: OwnedTransactionId, event_id: OwnedEventId) {
```

Comment:
It would be good to have some unit tests for this.

Reply:
Yep, will write a bunch in a subsequent PR!